### PR TITLE
[jit] Export BatchNorm functional and module, add necessary JIT support

### PR DESCRIPTION
**Pull request by wanchaol**
This PR did three things:

1. It export the BatchNorm functional and module, and rewrite some of the components to stay align with the current supported JIT features
2. In the process of export, add necessary compiler support for in_place op aug assign
4. change the test_jit behavior in add_module_test to utilize a single rng state during module initialization

### CODE REVIEW

File: torch/nn/modules/batchnorm.py [hunk 1]
```diff
@@ -3,12 +3,15 @@
...
 from .. import init
+from ..._jit_internal import weak_module, weak_script_method
 
 
 # TODO: check contiguous in THNN
 # TODO: use separate backend functions?
+@weak_module
 class _BatchNorm(Module):
     _version = 2
+    __constants__ = ['training', 'track_running_stats', 'momentum']
```
Comment (soumith) [inline torch/nn/modules/batchnorm.py]: if `training` is a constant, then `model.eval()` will fail right?
Another reported issue about `training` being constant (for dropout): https://github.com/pytorch/pytorch/issues/13991

Comment (wanchaol) [reply torch/nn/modules/batchnorm.py]: Yes this is a known issue for us, I also documented this issue in #12894, I think we still don't hav e a good solution yet, that's why we go with the weak_module approach to not interfere with users calling from python.

Comment (wanchaol) [reply torch/nn/modules/batchnorm.py]: @soumith to be clear on this, this will only fail when users using JIT and making a call from `ScriptModule` then want to set `model.eval()`, it will not affect the python behaviors.

Comment (soumith) [reply torch/nn/modules/batchnorm.py]: i see. thanks for the clarification.

So you are saying `model.eval()` will work on a ScriptModule correctly?
Or are you saying `model.eval()` on a regular nn.Module will continue to work correctly as before?

Comment (wanchaol) [reply torch/nn/modules/batchnorm.py]: I mean the second one :). 

To make `model.eval()` work on `ScriptModule` requires more changes on our current infrastructure, and we might want to work on this after the standard library export.

File: test/test_jit.py [hunk 2]
```diff
@@ -9972,15 +9993,20 @@ def __init__(self):
...
             # Check there are no Python ops by exporting
             self.assertExportImportModule(module, tensors)
-            create_module.last_graph = module.graph
+            create_script_module.last_graph = module.graph
+            return module(*args)
+
+        # Construct a normal nn module to stay consistent with create_script_module
+        # and make use of a single global rng_state in module initialization
+        def create_nn_module(*args, **kwargs):
+            module = nn_module(*constructor_args)
```
Comment (driazati) [inline test/test_jit.py]: Why is this necessary? It's constructing the module twice (here and line 9967)

File: test/test_jit.py [hunk 3]
```diff
@@ -9972,15 +9993,20 @@ def __init__(self):
 
             # Check there are no Python ops by exporting
             self.assertExportImportModule(module, tensors)
-            create_module.last_graph = module.graph
+            create_script_module.last_graph = module.graph
+            return module(*args)
+
+        # Construct a normal nn module to stay consistent with create_script_module
+        # and make use of a single global rng_state in module initialization
+        def create_nn_module(*args, **kwargs):
```
Comment (driazati) [inline test/test_jit.py]: Why is this necessary here?

File: torch/csrc/jit/script/compiler.cpp [hunk 4]
```diff
@@ -1424,6 +1427,31 @@ struct to_ir {
     }
   }
 
+  void emitAugAssignmentToSelectVar(const AugAssign& stmt) {
```
Comment (driazati) [inline torch/csrc/jit/script/compiler.cpp]: Could you add a comment with an example of when this is called?

File: test/test_jit.py (hunk 3 above)
Comment (wanchaol) [reply test/test_jit.py]: Consider a normal nn module with some random generator functions in its __init__ method: Taking batchnorm as a example, __init__ is calling `reset_parameters`, which have random distribution generated, if we initialize the normal BatchNorm module outside, it will generate a distribution base on a different rng state with ScriptModule, then forward pass will be wrong because it use a different random parameter. So we need to ensure normal nn module and scriptModule share the same rng state even during model initialization. 

This change make every normal nn module and scriptModule create and run only in `self.runAndSaveRng` to share the single state.

File: torch/csrc/jit/script/compiler.cpp (hunk 4 above)
Comment (wanchaol) [reply torch/csrc/jit/script/compiler.cpp]: added.

**Review by driazati — APPROVED** [review]
